File: src/api/k8s/customResources.ts
import { invoke } from '@tauri-apps/api/core';
import { listen, type UnlistenFn } from '@tauri-apps/api/event';
import { EventType, EventHandler } from '@/types/k8sEvent';
import { K8sResponse } from '@/types/k8sResponse';

export type CrdIdentity = {
  group: string;
  version: string;
  kind: string;
  plural: string;
  isNamespaced: boolean;
};

export interface CustomResourceEvent {
  type: EventType;
  object: Record<string, any>;
}

export async function createCustomResource({
  name,
  namespace,
  manifest,
  crd,
}: {
  name: string;
  namespace?: string;
  manifest: Record<string, any>;
  crd: CrdIdentity;
}): Promise<Record<string, any>> {
  const { group, version, kind, plural, isNamespaced } = crd;
  return await invoke<Record<string, any>>('create_custom_resource', {
    name,
    namespace,
    group,
    version,
    kind,
    plural,
    isNamespaced,
    manifest,
  });
}

export async function updateCustomResource({
  name,
  namespace,
  manifest,
  crd,
}: {
  name: string;
  namespace?: string;
  manifest: Record<string, any>;
  crd: CrdIdentity;
}): Promise<Record<string, any>> {
  const { group, version, kind, plural, isNamespaced } = crd;
  return await invoke<Record<string, any>>('update_custom_resource', {
    name,
    namespace,
    group,
    version,
    kind,
    plural,
    isNamespaced,
    manifest,
  });
}

export async function listCustomResources({
  name,
  namespaces,
  crd,
}: {
  name: string;
  namespaces?: string[];
  crd: CrdIdentity;
}): Promise<Record<string, any>[]> {
  const { group, version, kind, plural, isNamespaced } = crd;
  return await invoke<Record<string, any>[]>('list_custom_resources', {
    name,
    namespaces,
    group,
    version,
    kind,
    plural,
    isNamespaced,
  });
}

export async function watchCustomResources({
  name,
  namespaces,
  crd,
  onEvent,
}: {
  name: string;
  namespaces?: string[];
  crd: CrdIdentity;
  onEvent?: EventHandler<CustomResourceEvent>;
}): Promise<{ eventName: string; unlisten: UnlistenFn }> {
  const { group, version, kind, plural, isNamespaced } = crd;
  const eventName = await invoke<string>('watch_custom_resources', {
    name,
    namespaces,
    group,
    version,
    kind,
    plural,
    isNamespaced,
  });
  const unlisten = await listen<CustomResourceEvent>(eventName, (evt) => {
    try {
      onEvent?.(evt.payload);
    } catch (err) {
      console.error('Error in onEvent handler:', err);
    }
  });
  return { eventName, unlisten };
}

export async function deleteCustomResources({
  name,
  namespace,
  resourceNames,
  crd,
}: {
  name: string;
  namespace?: string;
  resourceNames: string[];
  crd: CrdIdentity;
}): Promise<K8sResponse[]> {
  const { group, version, kind, plural, isNamespaced } = crd;
  return await invoke<K8sResponse[]>('delete_custom_resources', {
    name,
    namespace,
    group,
    version,
    kind,
    plural,
    isNamespaced,
    resourceNames,
  });
}

File: src-tauri/src/commands/custom_resources.rs
//! Custom resource (CRD instance) commands.
//!
//! Not macro-generated: every call carries the GVK + plural + scope at runtime,
//! since the kind is not known at compile time.

use crate::commands::common::start_watch;
use crate::error::AppResult;
use crate::manager::k8s::dynamic_resources::DynamicK8sResources;
use crate::utils::watcher::WatchManager;
use serde_json::Value;
use tauri::AppHandle;

#[tauri::command]
pub async fn create_custom_resource(
    name: String,
    namespace: Option<String>,
    group: String,
    version: String,
    kind: String,
    plural: String,
    is_namespaced: bool,
    manifest: Value,
) -> AppResult<Value> {
    DynamicK8sResources::create(
        name,
        namespace,
        group,
        version,
        kind,
        plural,
        is_namespaced,
        manifest,
    )
    .await
}

#[tauri::command]
pub async fn update_custom_resource(
    name: String,
    namespace: Option<String>,
    group: String,
    version: String,
    kind: String,
    plural: String,
    is_namespaced: bool,
    manifest: Value,
) -> AppResult<Value> {
    DynamicK8sResources::update(
        name,
        namespace,
        group,
        version,
        kind,
        plural,
        is_namespaced,
        manifest,
    )
    .await
}

#[tauri::command]
pub async fn list_custom_resources(
    name: String,
    namespaces: Option<Vec<String>>,
    group: String,
    version: String,
    kind: String,
    plural: String,
    is_namespaced: bool,
) -> AppResult<Vec<Value>> {
    DynamicK8sResources::list(
        name,
        namespaces,
        group,
        version,
        kind,
        plural,
        is_namespaced,
    )
    .await
}

#[tauri::command]
#[allow(clippy::too_many_arguments)]
pub async fn watch_custom_resources(
    app_handle: AppHandle,
    name: String,
    namespaces: Option<Vec<String>>,
    group: String,
    version: String,
    kind: String,
    plural: String,
    is_namespaced: bool,
    state: tauri::State<'_, WatchManager>,
) -> AppResult<String> {
    // Multi-segment channel key. `validate_resource_name` permits slashes for
    // exactly this case, and `cluster_prefix` only inspects the first segment, so
    // per-cluster teardown still matches these watches.
    let resource_key = if group.is_empty() {
        format!("custom_resources/{version}/{plural}")
    } else {
        format!("custom_resources/{group}/{version}/{plural}")
    };

    start_watch(
        app_handle,
        name,
        &resource_key,
        namespaces,
        state,
        // `FnOnce`, so the GVK strings move in rather than being cloned per call.
        move |app_handle, ctx_name, ns_list, event_name, cancel| {
            DynamicK8sResources::watch(
                app_handle,
                ctx_name,
                ns_list,
                group,
                version,
                kind,
                plural,
                is_namespaced,
                event_name,
                cancel,
            )
        },
    )
    .await
}

#[tauri::command]
#[allow(clippy::too_many_arguments)]
pub async fn delete_custom_resources(
    name: String,
    namespace: Option<String>,
    group: String,
    version: String,
    kind: String,
    plural: String,
    is_namespaced: bool,
    resource_names: Vec<String>,
) -> AppResult<Vec<Result<String, String>>> {
    DynamicK8sResources::delete(
        name,
        namespace,
        group,
        version,
        kind,
        plural,
        is_namespaced,
        resource_names,
    )
    .await
}
